morebits.js

```diff
@@ -4566,215 +4577,1725 @@ Morebits.wiki.page = function(pageName, status) {
+				ctx.onGroupsFailure(this);
+				return;
+			}
+			query.add = ctx.addGroups;
+		}
+		if (ctx.removeGroups) {
+			query.remove = ctx.removeGroups;
+		}
+
+
+		ctx.groupsApi = new Morebits.wiki.api('changing user groups...', query, fnGroupsSuccess, ctx.statusElement, fnGroupsError);
+		ctx.groupsApi.setParent(this);
+		ctx.groupsApi.post();
+
+	};
+
+	/**
+	 * Notify a user via {@link Morebits.wiki.page}.  Main advantages are
+	 * ability to skip notifying bots or indefinitely sitewide-blocked
+	 * users, or users with specific template or optout links.  Some
+	 * options are customizable, but implies `setCreateOption('recreate')`
+	 * and `setFollowRedirect(true, false)`; other options are available
+	 * via `setPageobjectFunctions`.  Makes use of:
+	 * - {@link Morebits.wiki.user#setMessage|setMessage}
+	 * - {@link Morebits.wiki.user#setSectionTitle|setSectionTitle}
+	 * - {@link Morebits.wiki.user#setNotifyBots|setNotifyBots}
+	 * - {@link Morebits.wiki.user#setNotifyIndef|setNotifyIndef}
+	 * - {@link Morebits.wiki.user#setNotifySelf|setNotifySelf}
+	 * - {@link Morebits.wiki.user#setNotifySkips|setNotifySkips}
+	 * - {@link Morebits.wiki.user#setPageobjectFunctions|setPageobjectFunctions}
+	 *
+	 * The actual processing is handled in `fnProcessNotify`.
+	 *
+	 * @param {Function} [onSuccess] - Callback function to run on success.
+	 * @param {Function} [onFailure] - Callback function to run on failure.
+	 */
+	this.notify = function(onSuccess, onFailure) {
+		ctx.onNotifySuccess = onSuccess;
+		ctx.onNotifyFailure = onFailure || emptyFunction;
+
+		if (ctx.isIPRange) {
+			ctx.statusElement.error('Cannot notify IP ranges');
+			ctx.onNotifyFailure(this);
+			return;
+		}
+		// Check underscores
+		if (ctx.notifySelf && ctx.userName === mw.config.get('wgUserName')) {
+			ctx.statusElement.error('Skipping self notification');
+			ctx.onNotifyFailure(this);
+			return;
+		}
+
+		// Ensure user is loaded
+		if (fnDontNeedLoad('notify')) {
+			fnProcessNotify.call(this);
+		} else {
+			this.load(fnProcessNotify, ctx.onNotifyFailure);
+		}
+	};
+
+	// Send the notification
+	var fnProcessNotify = function() {
+		// Empty reason, message, and token handled by Morebits.wiki.page
+		if (!ctx.exists) {
+			ctx.statusElement.error('Cannot notify the user because the user does not exist');
+			ctx.onNotifyFailure(this);
+			return;
+		}
+
+		if (ctx.notifySkipTemplates && ctx.notifySkipTemplates.length && ctx.talkTemplates && ctx.talkTemplates.length) {
+			// More efficient to do a for loop, but this is prettier?
+			var tlDups = Morebits.array.dups(ctx.talkTemplates.concat(ctx.notifySkipTemplates));
+			if (tlDups.length) {
+				ctx.statusElement.error('User talk page transcludes ' + tlDups[0] + ', aborting notification');
+				ctx.onNotifyFailure(this);
+				return;
+			}
+		} else if (ctx.notifySkipLink && ctx.talkLinks && ctx.talkLinks.length) {
+			// Should be without leading protocol; relying on mw.Uri could help
```

Comment:
Is it worth leaning on `mw.Uri` just for this ?  I'm not sold, but it could simplify some possible user errors.

Reply:
mw.Uri is loaded by default on all pages, so there's no harm in using it. (Though I'm not sure why/how it's helpful here.)